cmd/devicesimulator/main.go

```diff
@@ -45,34 +69,42 @@ func main() {
 	agents := make([]*agent.Agent, *numDevices)
 	for i := 0; i < *numDevices; i++ {
 		agentName := fmt.Sprintf("device-%04d", i)
-		agentDir := filepath.Join(*certDir, agentName)
+		certDir := filepath.Join(agentConfigTemplate.ConfigDir, "certs")
+		agentDir := filepath.Join(*dataDir, agentName)
 		for _, filename := range []string{"ca.crt", "client-enrollment.crt", "client-enrollment.key"} {
-			if err := copyFile(filepath.Join(*certDir, filename), filepath.Join(agentDir, filename)); err != nil {
+			if err := copyFile(filepath.Join(certDir, filename), filepath.Join(agentDir, filename)); err != nil {
 				log.Fatalf("copying %s: %v", filename, err)
 			}
 		}
-		enrollmentUIEndpoint := ""
-
-		cfg := agent.Config{
-			ManagementEndpoint:   *managementEndpoint,
-			EnrollmentEndpoint:   *managementEndpoint,
-			EnrollmentUIEndpoint: enrollmentUIEndpoint,
-			DataDir:              agentDir,
-			Key:                  filepath.Join(agentDir, agent.KeyFile),
-			Cacert:               filepath.Join(agentDir, agent.CacertFile),
-			GeneratedCert:        filepath.Join(agentDir, agent.GeneratedCertFile),
-			EnrollmentCertFile:   filepath.Join(agentDir, agent.EnrollmentCertFile),
-			EnrollmentKeyFile:    filepath.Join(agentDir, agent.EnrollmentKeyFile),
-			TPMPath:              *tpmPath,
-			SpecFetchInterval:    util.Duration(*specFetchInterval),
-			StatusUpdateInterval: util.Duration(*statusUpdateInterval),
+
+		cfg := agent.NewDefault()
+		cfg.ConfigDir = agentConfigTemplate.ConfigDir
+		cfg.DataDir = agentDir
+		cfg.EnrollmentService = agentConfigTemplate.EnrollmentService
+		cfg.ManagementService = agent.ManagementService{
+			Config: client.Config{
+				Service: agentConfigTemplate.ManagementService.Config.Service,
+				AuthInfo: client.AuthInfo{
+					ClientCertificate:     filepath.Join(agentDir, agent.GeneratedCertFile),
```

Comment:
Shall we leave this AuthInfo empty, and let the agent startup fill it up if no ClientCertData and ClientKeyData, etc.. provided?

In here: https://github.com/flightctl/flightctl/pull/236/commits/125d8dcb622d66fbe5b97407cdf10a0285034d86#diff-f48445f706f26bd59d7bae605c4445d3c178237d2489223bf9c1a813417d9713R76

In such a way that you don't need to fill it up in the test harness either.

Reply:
Ah, good idea. Should that then also include the initialisation from `internal/agent/config.go` new lines 161-170?

Reply:
Hmm right, we could wrap all the logic in Complete, but then call complete from the agent Run, WDYT?

Reply:
This would break the pattern where we (1) read the user's config, (2) Complete() the fields that the user didn't specify or whose config depends on the user's configuration, and (3) Validate() the then complete and final configuration. Validate() also checks that files exist and can be opened, which is why I can only set the paths to these files after.

Ok, I'll move only the part of completing the management service config into the agent, as anyway we may want to have the URL returned by the enrollment service.

Reply:
done